lib/resources/virtualization.rb

```diff
@@ -0,0 +1,246 @@
+# encoding: utf-8
+# author: Takaaki Furukawa
+
+require 'hashie/mash'
+
+module Inspec::Resources
+  class Virtualization < Inspec.resource(1) # rubocop:disable Metrics/ClassLength
+    name 'virtualization'
+    desc 'Use the virtualization InSpec audit resource to test the virtualization platform on which the system is running'
+    example "
+      describe virtualization.system do
+        it { should eq 'docker' }
+      end
+
+      describe virtualization do
+        its('role') { should eq 'guest' }
+      end
+
+      describe virtualization.systems[:docker] do
+        it { should eq 'guest' }
+      end
+
+      if virtualization.system == 'docker'
+        describe file('/var/tmp/foo') do
+          it { should be_file }
+        end
+      end
+    "
+
+    def initialize
+      unless inspec.os.linux?
+        skip_resource 'The `virtualization` resource is not supported on your OS yet.'
+      end
+    end
+
+    def [](name)
+      # convert string to symbol
+      name = name.to_sym if name.is_a? String
+      data = collect_data_linux
+      data[name]
+    end
+
+    # add helper methods for easy access of properties
+    # allows users to use virtualization.role, virtualization.system, virtualization.systems
+    %w{role system systems}.each do |property|
+      define_method(property.to_sym) do
+        data = collect_data_linux
+        data[property.to_sym]
+      end
+    end
+
+    def params
+      collect_data_linux
+    end
+
+    def to_s
+      'Virtualization Detection'
+    end
+
+    private
+
+    def lxc_version_exists?
+      inspec.command('lxc-version').exist?
+    end
+
+    def docker_exists?
+      inspec.command('docker').exist?
+    end
+
+    def nova_exists?
+      inspec.command('nova').exist?
+    end
+
+    def collect_data_linux # rubocop:disable Metrics/MethodLength, Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity, Metrics/AbcSize
+      virtualization = Hashie::Mash.new unless virtualization
```

Comment:
Since `collect_data_linux` is used by each of the three defined accessor methods, it could be called multiple times in a single test (i.e. the user had multiple `its` statements). Should we cache the data in an instance variable and then return it if it's defined? That would eliminate multiple runs through this method.